File: ShareX_windows/ShareX_windows/Helpers/HeartSocket.cs
using System.Diagnostics;
using System.Net.Sockets;
using System.Threading;
using System.Threading.Tasks;

namespace ShareX_windows.Helpers
{
    public class HeartSocket
    {
        private SocketHelper _mainSocket;

        private static HeartSocket instance = null;
        public static HeartSocket Instance
        {
            get
            {
                if (instance == null)
                {
                    instance = new HeartSocket();
                }
                return instance;
            }
        }
        private static SocketHelper heartSocket;

        private HeartSocket()
        {
            heartSocket = new(1233);
        }
        System.Action disconnectAllSockets;
        public void initializeDisconnection(System.Action das)
        {
            disconnectAllSockets = das;
        }

        public async Task connect()
        {
            await heartSocket.hostServer();
            new Thread(async () =>
             {
                 var n = 0;
                 while (true)
                 {
                     try
                     {
                         Thread.Sleep(1000);
                         //var b = new byte[1] { 2 };
                         //heartSocket.s.Send(b, 1, System.Net.Sockets.SocketFlags.None);
                         //Debug.WriteLine($"heart socket read msg is {IsConnected(heartSocket.s)}");
                         //Debug.WriteLine($"heart socket read msg is {heartSocket.s.Receive(b, 1, System.Net.Sockets.SocketFlags.None)}");
                         //await heartSocket.send_msg(n.ToString());
                         //n++;
                         //Debug.WriteLine("heartSocket msg sent");
                         //await heartSocket.receiveMsg();
                         //Debug.WriteLine("heartSocket msg sent");
                         //Debug.WriteLine($"heart socket read msg is {await heartSocket.receiveMsg()}");
                         var heartSocketConnected = IsConnected(heartSocket.s);
                         Debug.WriteLine($"is connected {heartSocketConnected}");
                         if (!heartSocketConnected)
                         {
                             Debug.WriteLine($"heart socket disconnected");
                             await MainSocket.Instance.connect();
                             await connect();

                             disconnectAllSockets();
                         }


                     }
                     catch (SocketException e)
                     {
                         Debug.WriteLine($"heart Socket error is {e.Message}");
                         await MainSocket.Instance.connect();
                         disconnectAllSockets();

                     }

                 }
             }).Start();
        }

        public bool IsConnected(Socket socket)
        {
            try
            {
                bool part1 = socket.Poll(1000, SelectMode.SelectRead);
                bool part2 = (socket.Available == 0);
                if (part1 && part2)
                    return false;
                else
                    return true;
            }
            catch (SocketException) { return false; }
        }
    }
}

File: ShareX_android/app/src/main/java/com/example/sharex/helpers/HeartSocket.kt
package com.example.sharex.helpers

import android.util.Log
import com.example.sharex.MainActivity
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.net.Socket
import java.net.SocketException

object HeartSocket {
    var heartSocket = SocketHelper(1233)
    var isConnected = false
    public fun connect(ip : String, main:MainActivity){
        CoroutineScope(Dispatchers.IO).launch {
            while (!isConnected)
            {
                Log.d("heartSocket", "waiting for connection")
                isConnected = heartSocket.connectToServer(ip)
                if (isConnected)
                    Log.d("heartSocket", "heartSocket connected")
                else
                    Log.d("heartSocket", "heartSocket not connected")
            }

            while (true)
            {
                try {
                    Thread.sleep(1000)
//                    var b = ByteArray(1)
//                    Log.d("hearSocket", " is alive : ${heartSocket.inputStream.read(b)}")
//                    Log.d("hearSocket", " is alive 1 : ${heartSocket.inputStream.available()}")
//                    heartSocket.outputStream.write(byteArrayOf(1),0,1);
//                    val rcvMsg = heartSocket.receiveMsg();
//                    Log.d("hearSocket", "rcv msg is $rcvMsg ")
//                    heartSocket.sendMsg(rcvMsg)
//                    Log.d("hearSocket", "msg sent")
                    Log.d("hearSocket", "is connected = ${checkConnection(heartSocket.server)}")

                }
                catch (e : SocketException)
                {
                    isConnected = false
                    Log.d("hearSocket", "hearSocket error: ${e.message} ")
                    break
                }

            }
            main.disconnectPc()
            main.connectToPC(ip)
        }


    }
    suspend fun checkConnection(socket : Socket): Boolean{
        return socket.getInputStream().read() != -1
    }
}